src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NETSTANDARD2_0
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.CustomMessageHandler;
+
+[ExportWorkspaceService(typeof(ICustomMessageHandlerService)), Shared]
+internal sealed class CustomMessageHandlerService : ICustomMessageHandlerService, IDisposable
+{
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly Dictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    /// <summary>
+    /// Handlers of document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, ICustomMessageDocumentHandlerWrapper> _documentHandlers = new();
+
+    /// <summary>
+    /// Handlers of non-document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, ICustomMessageHandlerWrapper> _handlers = new();
+
+    private readonly ICustomMessageHandlerFactory _customMessageHandlerFactory;
+
+    private readonly object _lockObject = new();
+
+    private readonly AssemblyLoadContext _defaultLoadContext
+        = AssemblyLoadContext.GetLoadContext(typeof(CustomMessageHandlerService).Assembly)
+        ?? throw new InvalidOperationException($"Cannot get assembly load context for {nameof(CustomMessageHandlerService)}.");
```

Comment:
so i don't know enough about what is going on here.  but i would ideally like us to use IAnalyzerAssemblyLoaderProvider here, including CreateNewShadowCopyLoader.  This then handles all the load context stuff on our behalf.

This also then utilizes all the shadow copying, so we don't have to worry about locking dlls.  They can now change on disk, and we can be ok.

Reply:
I don't see anything in `IAnalyzerAssemblyLoaderInternal` that would allow me to unload the assembly load context used to load the assemblies. Is garbage collection going to take care of it at some point?

Reply:
> I don't see anything in IAnalyzerAssemblyLoaderInternal that would allow me to unload the assembly load context used to load the assemblies. Is garbage collection going to take care of it at some point?

1. Yes.
2. The idea of the IAnalyzerAssemblyLoaderInternal  is that you can get a shadow copy loader from it, from which you can load assemblies.  This can then be explicitly disposed when not needed, which will unload the ALCs.

Reply:
Done